### PR TITLE
nft bi directional translator

### LINKED ISSUE #7304: NFT bi-directional translator

### PR DESCRIPTION
**Pull request by povolev15**
bi-directional nft translator

Fixes #7304

### CODE REVIEW

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+public final class NftStateTranslator {
+
+    @NonNull
+    /**
+     * Converts a {@link com.hedera.node.app.service.mono.state.merkle.MerkleUniqueToken} and to {@link Nft}.
+     * @param merkleUniqueToken the {@link com.hedera.node.app.service.mono.state.merkle.MerkleUniqueToken}
+     * @return the {@link Nft} converted from the {@link com.hedera.node.app.service.mono.state.merkle.MerkleUniqueToken}
+     */
+    public static Nft accountFromMerkle(
```
Comment (tinker-michaelj) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java:45]: `nftFromMerkle()`?

Comment (povolev15) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java]: done resolved

**Comment by mhess-swl** [comment]
Is this PR based on the same branch as #7366? I'm seeing the same issues in both places

**Review by tinker-michaelj — DISMISSED** [review]
LGTM!

**Comment by jsync-swirlds** [comment]
> Is this PR based on the same branch as #7366? I'm seeing the same issues in both places

It is similar, as both are adapting the ID changes in modularized code for mono service.

**Comment by povolev15** [comment]
> > Is this PR based on the same branch as #7366? I'm seeing the same issues in both places
> 
> It is similar, as both are adapting the ID changes in modularized code for mono service.

@jsync-swirlds - I rebase on her pr

**Comment by povolev15** [comment]
in order not to have conflicts

**Review by tinker-michaelj — DISMISSED** [review]
LGTM

**Comment by jsync-swirlds** [comment]
> > > Is this PR based on the same branch as #7366? I'm seeing the same issues in both places
> > 
> > 
> > It is similar, as both are adapting the ID changes in modularized code for mono service.
> 
> @jsync-swirlds - I rebase on her pr

You'll also need to rebase to develop after that PR merges, which (hopefully) should resolve some of the build failures.

File: settings.gradle.kts [hunk 2]
```diff
@@ -103,7 +103,7 @@ gradleEnterprise {
 // The HAPI API version to use for Protobuf sources. This can be a tag or branch
 // name from the hedera-protobufs GIT repo.
 val hapiProtoVersion = "0.40.0-blocks-state-SNAPSHOT"
-val hapiProtoBranchOrTag = "7242-update-common-ids" // hapiProtoVersion
+val hapiProtoBranchOrTag = "add-pbj-types-for-state" // hapiProtoVersion
```
Comment (kimbor) [inline settings.gradle.kts:106]: Iris's PR #7366 also makes this change

**Review by kimbor — DISMISSED** [review]

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        final var nftPrevIdPair = merkleUniqueToken.getPrev();
+        builder.ownerPreviousNftId(merkelUniqueTokenToUniqueTokenId(nftPrevIdPair));
+
+        final var nftNextIdPair = merkleUniqueToken.getNext();
+        builder.ownerNextNftId(merkelUniqueTokenToUniqueTokenId(nftNextIdPair));
+
+        return builder.build();
+    }
+
+    private static @NonNull UniqueTokenId merkelUniqueTokenToUniqueTokenId(@NotNull NftNumPair merkleUniqueToken) {
```
Comment (qnswirlds) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java:74]:
```suggestion
    private static @NonNull UniqueTokenId merkelUniqueTokenToUniqueTokenId(@NonNull NftNumPair merkleUniqueToken) {
```

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+import com.hedera.hapi.node.state.token.Nft;
+import com.hedera.node.app.service.mono.state.merkle.internals.BitPackUtils;
+import com.hedera.node.app.service.mono.state.submerkle.EntityId;
+import com.hedera.node.app.service.mono.utils.EntityNum;
+import com.hedera.node.app.service.mono.utils.EntityNumPair;
+import com.hedera.node.app.service.mono.utils.NftNumPair;
+import com.hedera.node.app.service.token.ReadableNftStore;
+import com.hedera.pbj.runtime.io.buffer.Bytes;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import org.jetbrains.annotations.NotNull;
```
Comment (qnswirlds) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java:34]:
```suggestion
```

File: hedera-node/hedera-mono-service/src/main/java/module-info.java [hunk 5]
```diff
@@ -294,4 +294,5 @@
     requires org.slf4j;
     requires static com.github.spotbugs.annotations;
     requires com.hedera.node.app.service.token;
+    requires org.jetbrains.annotations;
```
Comment (qnswirlds) [inline hedera-node/hedera-mono-service/src/main/java/module-info.java:297]:
```suggestion
```

**Review by qnswirlds — CHANGES_REQUESTED** [review]
Use `@NonNull`, do not use `@NotNull`, no change needed in `module-info`.

File: hedera-node/hedera-mono-service/src/test/java/com/hedera/node/app/service/mono/state/merkle/MerkleUniqueTokenSerdeTest.java [hunk 6]
```diff
@@ -52,7 +53,9 @@ protected MerkleUniqueToken getExpectedObject(
             seededObject.setNext(NftNumPair.MISSING_NFT_NUM_PAIR);
             return seededObject;
         } else {
-            return seededObject;
+            final var pbjNft = NftStateTranslator.accountFromMerkle(seededObject);
+            final var merkleUniqueToke = NftStateTranslator.merkleUniqueTokenFromNft(pbjNft);
```
[the remark this answers is not in the document]
Comment (povolev15) [reply hedera-node/hedera-mono-service/src/test/java/com/hedera/node/app/service/mono/state/merkle/MerkleUniqueTokenSerdeTest.java:57]: done thanks

File: settings.gradle.kts (hunk 2 above)
Comment (povolev15) [reply settings.gradle.kts]: I just rebased

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java (hunk 4 above)
Comment (jsync-swirlds) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java:34]: It is possible to set IntelliJ to exclude this import so it doesn't help you put the wrong annotation in.

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java (hunk 3 above)
Comment (povolev15) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java:74]: done thanks

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java (hunk 4 above)
Comment (povolev15) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java:34]: removed

File: hedera-node/hedera-mono-service/src/main/java/module-info.java (hunk 5 above)
Comment (povolev15) [reply hedera-node/hedera-mono-service/src/main/java/module-info.java:297]: done thanks

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+    public static Nft nftFromMerkleUniqueToken(
+            @NonNull final com.hedera.node.app.service.mono.state.merkle.MerkleUniqueToken merkleUniqueToken) {
+        requireNonNull(merkleUniqueToken);
+        var builder = Nft.newBuilder();
+        final var nftIdPair = merkleUniqueToken.getKey().asNftNumPair();
+        builder.id(merkelUniqueTokenToUniqueTokenId(nftIdPair));
+
+        builder.ownerId(AccountID.newBuilder()
+                        .accountNum(
+                                merkleUniqueToken.getOwner().toGrpcAccountId().getAccountNum())
```
Comment (jsync-swirlds) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java:53]: If we're converting GRPC account IDs, we should convert them all the way to hapi AccountID before starting the builder, and use the full account ID, not just the number.  This code currently throws out the shard and realm.

Comment (povolev15) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java:53]: correct there is another task for @iwsimon and @kimbor to add all as part of the process of the protobuf conversion from AccountId and TokenId

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java (hunk 4 above)
Comment (povolev15) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java:34]: I removed it as aprt of @qnswirlds comments

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+    @NonNull
+    /**
+     * Converts a {@link com.hedera.node.app.service.mono.state.merkle.MerkleUniqueToken} and to {@link Nft}.
+     * @param merkleUniqueToken the {@link com.hedera.node.app.service.mono.state.merkle.MerkleUniqueToken}
+     * @return the {@link Nft} converted from the {@link com.hedera.node.app.service.mono.state.merkle.MerkleUniqueToken}
+     */
+    public static Nft nftFromMerkleUniqueToken(
+            @NonNull final com.hedera.node.app.service.mono.state.merkle.MerkleUniqueToken merkleUniqueToken) {
+        requireNonNull(merkleUniqueToken);
+        var builder = Nft.newBuilder();
```
Comment (qnswirlds) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java:46]: nit:
```suggestion
        final var builder = Nft.newBuilder();
```

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+    @NonNull
+    /***
+     * Converts a {@link Nft} to a {@link com.hedera.node.app.service.mono.state.merkle.MerkleUniqueToken}.
+     * @param nft the {@link Nft} to convert
+     * @return the {@link com.hedera.node.app.service.mono.state.merkle.MerkleUniqueToken}
+     */
+    public static com.hedera.node.app.service.mono.state.merkle.MerkleUniqueToken merkleUniqueTokenFromNft(
+            @NonNull Nft nft) {
+        requireNonNull(nft);
+        com.hedera.node.app.service.mono.state.merkle.MerkleUniqueToken merkleUniqueToken =
```
Comment (qnswirlds) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java:110]: nit:
```suggestion
        final com.hedera.node.app.service.mono.state.merkle.MerkleUniqueToken merkleUniqueToken =
```

**Review by qnswirlds — DISMISSED** [review]

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+                new com.hedera.node.app.service.mono.state.merkle.MerkleUniqueToken();
+
+        if (nft.hasId()) {
+            merkleUniqueToken.setKey(EntityNumPair.fromNums(
+                    EntityNum.fromLong(nft.id().tokenId().tokenNum()),
+                    EntityNum.fromLong(nft.id().serialNumber())));
+        }
+        merkleUniqueToken.setOwner(EntityId.fromNum(nft.ownerId().accountNum()));
```
Comment (jsync-swirlds) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java:118]:
```suggestion
        merkleUniqueToken.setOwner(EntityId.fromGrpcAccountId(PbjConverter.fromPbj(nft.ownerId())));
```

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+        merkleUniqueToken.setSpender(EntityId.fromNum(nft.spenderId().accountNum()));
```
Comment (jsync-swirlds) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java:119]:
```suggestion
        merkleUniqueToken.setSpender(EntityId.fromGrpcAccountId(PbjConverter.fromPbj(nft.spenderId())));
```

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+        builder.id(merkelUniqueTokenToUniqueTokenId(nftIdPair));
+
+        builder.ownerId(AccountID.newBuilder()
+                        .accountNum(
+                                merkleUniqueToken.getOwner().toGrpcAccountId().getAccountNum())
+                        .build())
+                .spenderId(AccountID.newBuilder()
+                        .accountNum(
+                                merkleUniqueToken.getSpender().toGrpcAccountId().getAccountNum())
+                        .build())
```
Comment (jsync-swirlds) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java:57]:
```suggestion
                .spenderId(
                        PbjConverter.toPbj(merkleUniqueToken.getSpender().toGrpcAccountId()))
```

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java (hunk 7 above)
Comment (jsync-swirlds) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java:52]: Why defer?  These two lines are pretty easy to update.

**Review by kimbor — DISMISSED** [review]
lgtm!

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+            @NonNull final com.hedera.node.app.service.mono.state.merkle.MerkleUniqueToken merkleUniqueToken) {
+        requireNonNull(merkleUniqueToken);
+        var builder = Nft.newBuilder();
+        final var nftIdPair = merkleUniqueToken.getKey().asNftNumPair();
+        builder.id(merkelUniqueTokenToUniqueTokenId(nftIdPair));
+
+        builder.ownerId(AccountID.newBuilder()
+                        .accountNum(
+                                merkleUniqueToken.getOwner().toGrpcAccountId().getAccountNum())
+                        .build())
```
Comment (jsync-swirlds) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java:53]:
```suggestion
        builder.ownerId(AccountID.newBuilder()
                        .accountNum(
                                PbjConverter.toPbj(merkleUniqueToken.getOwner().toGrpcAccountId()))
```

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java (hunk 8 above)
Comment (povolev15) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java:46]: done thanks

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java (hunk 9 above)
Comment (povolev15) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java:110]: done thanks

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java (hunk 10 above)
Comment (povolev15) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java:118]: done thanks

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java (hunk 11 above)
Comment (povolev15) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java]: done thanks

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java (hunk 12 above)
Comment (povolev15) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java:58]: done thanks

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java (hunk 13 above)
Comment (povolev15) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/NftStateTranslator.java:54]: done thanks

**Review by jsync-swirlds — APPROVED** [review]

**Review by qnswirlds — APPROVED** [review]

**Review by Neeharika-Sompalli — APPROVED** [review]

**Pull request merged**